### PR TITLE
[transducers-js] Added comp overloads.

### LINKED ISSUE #19276: transducers-js new compile errors
Due to Microsoft/TypeScript#16368 we are detecting some new compile errors in the tests.
This library heavily uses generic types -- previously when comparing generic types we would squash all type parameters to `any`, making comparisons almost always succeed.
@colinkahn @dphilipson could you look at the new compile errors and see if the type definitions need to be changed?

### PR DESCRIPTION
**Pull request by NaridaL**
Added overloads for Transducer arguments, as those seem the most common.
Fixed N identical argument overload.
Fixes #19276 
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: #19276 
- [ ] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by NaridaL** [comment]
I'm having issues.. first `npm run lint` ran without issues locally. Now I'm getting `Cannot find module typescript`... I'll try again later.

File: types/transducers-js/index.d.ts [hunk 1]
```diff
@@ -39,8 +39,18 @@ export function isReduced(x: any): boolean;
...
+// 2-4 Transducers
+export function comp<A, B, C>(a: Transducer<A, B>, b: Transducer<B, C>): Transducer<A, C>;
+export function comp<A, B, C, D>(a: Transducer<A, B>, b: Transducer<B, C>, c: Transducer<C, D>): Transducer<A, D>;
+export function comp<A, B, C, D, E>(a: Transducer<A, B>, b: Transducer<B, C>, c: Transducer<C, D>, d: Transducer<D, E>): Transducer<A, E>;
+// N identical Transducers
+export function comp<A>(...args: Array<Transducer<A, A>>): Transducer<A, A>;
+// 2-4 arbitrary functions
+export function comp<A, B, C>(a: (a: A) => B, b: (b: B) => C): (a: A) => C;
+export function comp<A, B, C, D>(a: (a: A) => B, b: (b: B) => C, c: (c: C) => D): (a: A) => D;
+export function comp<A, B, C, D, E>(a: (a: A) => B, b: (b: B) => C, c: (c: C) => D, d: (d: D) => E): (a: A) => E;
```
Comment (dphilipson) [inline types/transducers-js/index.d.ts]: Aren't these non-transducer overloads backwards? `comp(f, g)(x)` is `f(g(x))`, meaning the output of `g` is the input of `f`.

Comment (NaridaL) [reply types/transducers-js/index.d.ts]: Yes... wouldn't the transducers overloads be backwards too then?

Comment (dphilipson) [reply types/transducers-js/index.d.ts]: Actually no. The definition of `Transducer` is
```ts
export type Transducer<TInput, TOutput> =
    <TResult, TCompleteResult>(
        xf: CompletingTransformer<TResult, TCompleteResult, TOutput>
    ) => CompletingTransformer<TResult, TCompleteResult, TInput>;
```
which stripping down to the relevant parts is
```ts
type Transducer<TInput, TOutput> = (xf: Thingy<TOutput>) => Thingy<TInput>
```
so the transducer type parameters are actually the opposite direction that they appear in the function.

File: types/transducers-js/index.d.ts [hunk 2]
```diff
@@ -13,10 +14,12 @@ export type Reducer<TResult, TInput> = (result: TResult, input: TInput) => TResu
 
 // Common case: Transducer<TInput, TOutput> =
 //   Transformer<TResult, TOutput> => Transformer<TResult, TInput>.
-export type Transducer<TInput, TOutput> =
-    <TResult, TCompleteResult>(
-        xf: CompletingTransformer<TResult, TCompleteResult, TOutput>
-    ) => CompletingTransformer<TResult, TCompleteResult, TInput>;
+export interface Transducer<TInput, TOutput> {
+    <TResult>(xf: Transformer<TResult, TOutput>)
```
Comment (NaridaL) [inline types/transducers-js/index.d.ts]: I'm not sure about this... the comment above indicates this should be allowed, and without it there's an error in the test... (but it's not directly related).

Comment (dphilipson) [reply types/transducers-js/index.d.ts]: Oh huh, I just checked out the code and you're right about a seemingly unrelated failure.

It looks like that line in the tests
```ts
const xf = t.comp(t.map(reverse), t.cat);
```
is a bit too much to ask from type inference. Without the overload, `xf` is inferred as `CompletingTransformer<{}, {}, number[]>`. Even with the overload which makes the error go a way, `xf` is inferred as the somewhat unsatisfying `Transducer<number[], any>`. I tried to make it work with stuff like
```ts
const xf = t.comp<number[], number[], number>(t.map(reverse), t.cat);
```
which led to other complicated errors.

It looks like this is indicative of a deeper problem that typical usage of this library simply asks too much of the type inferencer. I'm happy to add the overload, and we may just have to live with not getting satisfying types from this library when using `comp`.

(Shameless plug: these sorts of issues are exactly why I wrote [Transducist](https://github.com/dphilipson/transducist)).

**Comment by dphilipson** [comment]
Looks good to me! @NaridaL Do I have to do something else to update my review status?

And thanks for taking the time to do this. The typings in this library are no joke.

File: types/transducers-js/index.d.ts (hunk 2 above)
Comment (NaridaL) [reply types/transducers-js/index.d.ts]: Yeah, I figured something like that... https://github.com/DefinitelyTyped/DefinitelyTyped/issues/19276#issuecomment-325168002

Thanks for taking a look in any case!

Your lib looks useful! (I'm more a fan of inherent TS libs anyways :-D)

**Comment by NaridaL** [comment]
@dphilipson If you go to the "files changed" you can "review changes" --> "approve".

**Review by dphilipson — APPROVED** [review]

**Pull request merged**